File: src/122_Best_Time_to_Buy_and_Sell_Stock_II/main.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # 无限次数的股票买卖，那就是任何涨幅都加在一起
        minVal = prices[0]

        rtn = 0 # 又有点像滑动窗口
        for i in range(1, len(prices)):
            if prices[i] > minVal:
                rtn += (prices[i] - minVal)
                minVal = prices[i]
            else:
                minVal = prices[i]

        return rtn

if __name__ == "__main__":
    gua = Solution()
    # rtn = gua.isMatch(s = "aaa", p = "ab*ac*a")
    rtn = gua.maxProfit(prices=[7,1,5,3,6,4])
    print("expect result = 7, actual result = %d " % rtn )

    rtn = gua.maxProfit(prices=[7,6,4,3,1])
    print("expect result = 0, actual result = %d " % rtn )

    rtn = gua.maxProfit(prices=[1,2,3,4,5])
    print("expect result = 4, actual result = %d " % rtn )

File: src/122_Best_Time_to_Buy_and_Sell_Stock_II/main.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    //Runtime: 4 ms, faster than 90.87%
    //Memory Usage: 13.2 MB, less than 27.74%
    int maxProfit(vector<int>& prices) {
        // dp问题，拿到就想状态是啥，其中一个是天数，另一个根据题意，就设置为目前为止的卖出最大收益，还有一个是目前为止买入的最大收益
        int size = prices.size();
        vector<int> sell = vector<int>(size);
        vector<int> buyy = vector<int>(size);

        buyy[0] = -prices[0];
        for (int i = 1; i < size; ++i) {
            sell[i] = max(sell[i-1], prices[i]+ buyy[i-1]);
            buyy[i] = max(buyy[i-1], sell[i-1]- prices[i]);
        }
        return sell.back();
    }
};

int main() {
  Solution gua;
  int rtn;
  vector<int> prices;

  rtn = gua.maxProfit(prices = {7,1,5,3,6,4});
  cout << "expect value = 7, actual value = " << rtn << endl;

  rtn = gua.maxProfit(prices = {1,2,3,4,5});
  cout << "expect value = 4, actual value = " << rtn << endl;

  rtn = gua.maxProfit(prices = {7,6,4,3,1});
  cout << "expect value = 0, actual value = " << rtn << endl;

  return 0;
}

